### ISSUE TITLE
Order of opt.zero_grad(), loss.backward(), opt.step()

### ISSUE THREAD

**Issue by yw155** [vita-epfl/openpifpaf#123]
Hi @svenkreiss, I have a question about the following codes in the 'trainer.py':

        if loss is not None:
            loss.backward()
        if apply_gradients:
            self.optimizer.step()
            self.optimizer.zero_grad()
            self.step_ema()

The order of above functions is not the same as the usual order like: zero_grad, backward, step. I would like to ask you whether this change will affect the training or not. Thanks a lot.

**Comment by svenkreiss** [comment]
It's the same order as you say. Just starts with backward. But it is in a loop and there is a zero_grad before the loop.

It is done for correct behavior with apply_gradients. This allows to accumulate gradients across multiple badges.

**Comment by yw155** [comment]
I got that. Thanks for your reply.
